README.md
<img align="right" src="https://user-images.githubusercontent.com/225151/75162092-3ef2f780-571d-11ea-8d4e-616ccbd1c924.png" width="200" height="200"/>

`check_prometheus_metric`
=========================
![CI](https://github.com/magenta-aps/check_prometheus_metric/workflows/CI/badge.svg)

Nagios plugin for alerting on Prometheus query results.

# Usage
```
  check_prometheus_metric.sh - Nagios plugin for checking Prometheus metrics.
  
  Usage:
    check_prometheus_metric.sh -H HOST -q QUERY -w FLOAT[:FLOAT] -c FLOAT[:FLOAT]
                               -n NAME [-m METHOD] [-O] [-i] [-p]

  Options:
    -H HOST          URL of Prometheus host to query.
    -q QUERY         Prometheus query, in single quotes, that returns a float.
    -w FLOAT[:FLOAT] Warning level value (must be a float or nagios-interval).
    -c FLOAT[:FLOAT] Critical level value (must be a float or nagios-interval).
    -n NAME          A name for the metric being checked.
    -m METHOD        Comparison method, one of gt, ge, lt, le, eq, ne.
                     (Defaults to ge unless otherwise specified).
    -C CURL_OPTS     Additional flags to curl. Can be passed multiple times. 
                     Options and option values must be passed separately.
                     e.g. -C --connect-timeout -C 10 -C --cacert -C /path/to/ca.crt
    -O               Accept NaN as an "OK" result.
    -i               Print the extra metric information into the Nagios message.
    -p               Add perfdata to check output.

  Examples:
    check_prometheus_metric -q 'up{job=\"job_name\"}' -w :1 -c :1
    # Check that job is up. If not, critical.
    
    check_prometheus_metric -q 'node_load1' -w :0.05 -c :0.1
    # Check load is below 0.05 (warning) and 0.1 (critical).

    check_prometheus_metric -q 'go_threads' -w 15:25 -c :
    # Check thread count is between 15-25, warning if outside this interval.

  Dependencies:
    Requires bash, curl, cut, echo, grep, jq and sed to be in $PATH.
```
Note: `nagios-interval` refers to [this syntax](http://nagios-plugins.org/doc/guidelines.html#THRESHOLDFORMAT)

# Icinga configuration
To utilize the script with [Icinga 2](https://icinga.com/docs/icinga2/), four
steps must be taken;

1. The CheckCommand definition must be added to `/etc/icinga2/conf.d/check_prometheus_metric.conf`:
```
object CheckCommand "check_prometheus_metric" {
  import "plugin-check-command"
  command = [ "/usr/lib/nagios/plugins/check_prometheus_metric" ]

  arguments = {
        "-H" = {
                value = "$check_prometheus_metric_url$"
                description = "URL of Prometheus host to query."
        }
        "-q" = {
                value = "$check_prometheus_metric_query$"
                description = "Prometheus query, that returns a float."
        }
        "-w" = {
                value = "$check_prometheus_metric_warning$"
                description = "Warning level value (float or nagios-interval)."
        }
        "-c" = {
                value = "$check_prometheus_metric_critical$"
                description = "Critical level value (float or nagios-interval)."
        }
        "-n" = {
                value = "$check_prometheus_metric_name$"
                description = "A name for the metric being checked."
        }
    }
}
```
2. The checking script, must be added to `/usr/lib/nagios/plugins/check_prometheus_metric`:
```
GIT_REPO=https://github.com/magenta-aps/check_prometheus_metric
VERSION=$(curl -sL -H "Accept: application/json" ${GIT_REPO}/releases/latest | jq -r .tag_name)
curl -sL -o . ${GIT_REPO}/releases/download/${VERSION}/check_prometheus_metric.sh
chmod +x check_prometheus_metric.sh
sudo mv check_prometheus_metric.sh /usr/lib/nagios/plugins/check_prometheus_metric
```

3. Install the required software to run the script:
```
sudo apt-get update
sudo apt-get install -y bash coreutils curl grep jq sed
```

4. Add Service definitions for whatever is to be monitored:

`/etc/icinga2/conf.d/check_pi_service.conf`:
```
apply Service "pi" {
  import "generic-service"

  check_command = "check_prometheus_metric"

  vars.check_prometheus_metric_url = "prometheus:9090"
  vars.check_prometheus_metric_query = "pi"
  vars.check_prometheus_metric_warning = "3:4"
  vars.check_prometheus_metric_critical = "1:6"
  vars.check_prometheus_metric_name = "pi"
  
  command_endpoint = host.vars.client_endpoint
  assign where host.name == NodeName
}
```

# Nagios configuration
You need to add the following commands to your Nagios configuration to use it:
```
define command {
    command_name check_prometheus
    command_line $USER1$/check_prometheus_metric.sh -H '$ARG1$' -q '$ARG2$' -w '$ARG3$' -c '$ARG4$' -n '$ARG5$' -m '$ARG6$'
}
```


CHANGELOG
0.0.4
=====
Added CHANGELOG
Added logo in README.md
Added configuration documentation to README.md
Added additional preflight check for validation prometheus server connection.
Added test case against icinga itself.
Fixed a typo in the README.md file
Fixed bug causing wrong exit code for WARNING/CRITICAL returns

0.0.3
=====
Rework plugin to new structure.

Added nagios-interval syntax support
Added bats testing framework
- Added tests for current (old) behavior and syntax.
- Added tests for new behavior and syntax.
Autodetect vector / scalar query types based on query result.
Check intervals using jq instead of test -v
Support floats instead of integers all around.
Utilize epsilon instead of 1 for old syntax.
Added CI testing and release management.

0.0.2
=====
Rework project structure

0.0.1
=====
Created release workflow


test.sh
#!/bin/bash

NETWORK_NAME=nagios_plugins_testnetwork
PROMETHEUS_NAME=nagios_plugins_prometheus
PROMETHEUS_PORT=9090
PUSHGATEWAY_NAME=nagios_plugins_pushgateway
PUSHGATEWAY_PORT=9091
ICINGA_NAME=icinga2
ICINGA_PORT=5665

PLUGIN_SCRIPT=build/output.sh

export PLUGIN_SCRIPT=${PLUGIN_SCRIPT}
export PROMETHEUS_PORT=${PROMETHEUS_PORT}
export PUSHGATEWAY_PORT=${PUSHGATEWAY_PORT}
export ICINGA_PORT=5665

source tests/test_utils.bash

build_script

start_docker_network
start_prometheus
start_pushgateway
start_icinga

set_metric "pi" "3.14"

echo ""
echo "Waiting until prometheus sees pushed metric"
wait_for_metric "scalar(pi)" 3.14
echo ""

echo "Ordinary tests"
bats tests/*.bats

echo "Integration tests"
bats tests/icinga/*.bats


tests/test_utils.bash
function test_parameters_full() {
    local _PARAMETERS
    local _OUTPUT
    _PARAMETERS=$1
    _OUTPUT=$(bash ${PLUGIN_SCRIPT} -H "${PROMETHEUS_SERVER}" ${_PARAMETERS} -n "tc")
    printf '%s' "${_OUTPUT}"
    return 0
}

function test_parameters() {
    local _PARAMETERS
    local _OUTPUT
    _PARAMETERS=$1
    _OUTPUT=$(bash ${PLUGIN_SCRIPT} -H "${PROMETHEUS_SERVER}" ${_PARAMETERS} -n "tc" | tail -1)
    printf '%s' "${_OUTPUT}"
    return 0
}

function wait_for_metric() {
    local _RAW_RESULT
    local _PARSED_RESULT="NaN"
    local _EXPECTED_RESULT="$2"
    # Keep running query until data is found, i.e. non-nan returned
    until [ "${_PARSED_RESULT}" == "${_EXPECTED_RESULT}" ]; do
        printf '.'
        sleep 1
        _RAW_RESULT=$(curl -s --data-urlencode "query=$1" "http://localhost:${PROMETHEUS_PORT}/api/v1/query")
        _PARSED_RESULT=$(echo "${_RAW_RESULT}" | jq -r .data.result[1])
    done
    printf '%s' "${_PARSED_RESULT}"
    return 0
}

function set_metric() {
    echo "$1 $2" | curl -s --data-binary @- http://localhost:${PUSHGATEWAY_PORT}/metrics/job/constrants 1>/dev/null 2>/dev/null
}

function build_script() {
    echo ""
    echo "Compiling script"
    mkdir -p build/
    bash tools/compile.sh > ${PLUGIN_SCRIPT}
    chmod +x ${PLUGIN_SCRIPT}
}

function start_docker_network() {
    echo ""
    echo "Creating docker network"
    docker network rm ${NETWORK_NAME}
    docker network create ${NETWORK_NAME}
}

function start_prometheus() {
    echo ""
    echo "Creating configuration file"
    CONFIG_FILE=$(mktemp --suffix=.yml)
    chmod 666 ${CONFIG_FILE}
    echo "${CONFIG_FILE}"

    echo ""
    echo "Templating configuration file"
    sed -e "s/\${PUSHGATEWAY_NAME}/${PUSHGATEWAY_NAME}/" tests/configuration/prometheus.yml > ${CONFIG_FILE}

    echo ""
    echo "Starting prometheus container"
    docker rm -f ${PROMETHEUS_NAME}
    docker run --name ${PROMETHEUS_NAME} --net=${NETWORK_NAME} -d -p ${PROMETHEUS_PORT}:9090 -v ${CONFIG_FILE}:/etc/prometheus/prometheus.yml prom/prometheus

    echo ""
    echo "Waiting until prometheus is up"
    until $(curl --output /dev/null --silent --fail http://localhost:${PROMETHEUS_PORT}); do
        printf '.'
        sleep 1
    done

    echo ""
    echo "Waiting until prometheus sees itself"
    QUERY_SCALAR_UP="scalar(up{instance=\"localhost:9090\"})"
    wait_for_metric "${QUERY_SCALAR_UP}" 1
    echo ""
}

function start_pushgateway() {
    echo ""
    echo "Starting pushgateway container"
    docker rm -f ${PUSHGATEWAY_NAME}
    docker run --name ${PUSHGATEWAY_NAME} --net=${NETWORK_NAME} -d -p ${PUSHGATEWAY_PORT}:9091 prom/pushgateway

    echo ""
    echo "Waiting until pushgateway is up"
    until $(curl --output /dev/null --silent --fail http://localhost:${PUSHGATEWAY_PORT}); do
        printf '.'
        sleep 1
    done
}

function start_icinga() {
    echo "Cleaning up old container..."
    docker rm -f icinga

    echo "Creating icinga container..."
    docker create --name icinga -p 80:80 -p 5665:5665 -h icinga2 --net=${NETWORK_NAME} -t jordan/icinga2:latest

    echo "Transfering monitoring scripts..."
    docker cp ${PLUGIN_SCRIPT} icinga:/usr/lib/nagios/plugins/check_prometheus_metric

    echo "Starting icinga..."
    docker start icinga

    echo "Waiting for icinga to generate files..."
    docker exec icinga /bin/bash -c "while ! [ -f /etc/icinga2/conf.d/services.conf ]; do printf '.'; sleep 1; done"
    echo ""

    echo "Transfering icinga configuration files..."
    docker cp tests/configuration/icinga.conf icinga:/etc/icinga2/conf.d/check_prometheus_metric.conf

    echo "Installing script dependencies..."
    docker exec icinga /bin/bash -c "apt-get update && apt-get install -y curl jq"

    echo "Fetching API password..."
    PASSWORD=$(docker exec icinga /bin/bash -c "cat /etc/icinga2/conf.d/api-users.conf | grep -o 'password = .*'| cut -f2 -d '\"'")
    echo "${PASSWORD}" 
    export "ICINGA_PASSWORD=${PASSWORD}"

    echo "Wait for icinga to be online..."
    # TODO: Actually check against the API
    sleep 10
}

function _fetch_service() {
    curl -s -k -u "root:${ICINGA_PASSWORD}" -H 'Accept: application/json' \
         -H 'X-HTTP-Method-Override: GET' \
         -X POST "https://localhost:${ICINGA_PORT}/v1/objects/services" \
         -d "{\"filter\": \"match(\\\"$1\\\",service.name)\"}"
}

function _order_recheck() {
    curl -s -k -u "root:${ICINGA_PASSWORD}" -H 'Accept: application/json' \
         -X POST "https://localhost:${ICINGA_PORT}/v1/actions/reschedule-check" \
         -d "{\"type\": \"Service\", \"filter\": \"service.name==\\\"$1\\\"\", \"force\": true}"
}

function recheck_service() {
    local _NOW
    local _RECHECK
    local _LAST_CHECKED
    local _KEEP_GOING="false"

    # Order a service update
    _RECHECK=$(_order_recheck $1)

    # Keep running until service is rechecked
    until [ "${_KEEP_GOING}" == "true" ]; do
        # Find out when we last checked the service
        _LAST_CHECKED=$(_fetch_service $1 | jq .[][0].attrs.last_check)
        _NOW=$(date +%s)
        # Keep going until last_checked is in the past
        _KEEP_GOING=$(echo "{\"last_checked\": ${_LAST_CHECKED}, \"now\": ${_NOW}}" | jq ".last_checked <= .now")
        printf '.'
        sleep 1
    done
}

function get_service_state() {
    _fetch_service $1 | jq .[][0].attrs.state
}


tests/test_config.bash

PLUGIN_SCRIPT=build/output.sh
PROMETHEUS_PORT=9090
PUSHGATEWAY_PORT=9091
ICINGA_PORT=5665

ICINGA_PASSWORD=2f6ef58b59247ddc


src/check_prometheus_metric.sh
#!/bin/bash
#
# check_prometheus_metric.sh - Nagios plugin wrapper for checking Prometheus
#                              metrics. Requires curl and jq to be in $PATH.

# Default configuration:
CURL_OPTS=()
COMPARISON_METHOD=ge
NAN_OK="false"
NAGIOS_INFO="false"
PERFDATA="false"
PROMETHEUS_QUERY_TYPE=""

# Constants
#----------
# Nagios status codes:
OK=0
WARNING=1
CRITICAL=2
UNKNOWN=3

# Variables:
NAGIOS_STATUS=UNKNOWN
NAGIOS_SHORT_TEXT='an unknown error occured'
NAGIOS_LONG_TEXT=''

# Includes
source preamble.bash
source parse.bash
source usage.bash

# Main

# Code:
function check_dependencies() {
    if ! [ -x "$(command -v curl)" ]; then
        NAGIOS_STATUS=UNKNOWN
        NAGIOS_SHORT_TEXT='missing "curl" command'
        exit
    fi

    if ! [ -x "$(command -v jq)" ]; then
        NAGIOS_STATUS=UNKNOWN
        NAGIOS_SHORT_TEXT='missing "jq" command'
    fi
}

function process_command_line {

  while getopts ':H:q:w:c:m:n:C:Oipt:' OPT "$@"
  do
    case ${OPT} in
      H)        PROMETHEUS_SERVER="$OPTARG" ;;
      q)        PROMETHEUS_QUERY="$OPTARG" ;;
      n)        METRIC_NAME="$OPTARG" ;;

      m)        # If invalid operator name
                if ! [[ ${OPTARG} =~ ^([lg][et]|eq|ne)$ ]]; then
                    NAGIOS_SHORT_TEXT="invalid comparison method: ${OPTARG}"
                    NAGIOS_LONG_TEXT="$(usage)"
                    exit
                fi
                COMPARISON_METHOD=${OPTARG}
                ;;

      c)        # If malformed
                if ! is_float_or_interval "${OPTARG}"; then
                  NAGIOS_SHORT_TEXT='-c CRITICAL_LEVEL requires a float or interval'
                  NAGIOS_LONG_TEXT="$(usage)"
                  exit
                fi
                CRITICAL_LEVEL=${OPTARG}
                ;;

      w)        # If malformed
                if ! is_float_or_interval "${OPTARG}"; then
                  NAGIOS_SHORT_TEXT='-w WARNING_LEVEL requires a float or interval'
                  NAGIOS_LONG_TEXT="$(usage)"
                  exit
                fi
                WARNING_LEVEL=${OPTARG}
                ;;

      C)        CURL_OPTS+=("${OPTARG}")
                ;;
      O)        NAN_OK="true"
                ;;

      i)        NAGIOS_INFO="true"
                ;;

      p)        PERFDATA="true"
                ;;

      t)
                NAGIOS_LONG_TEXT+="Note: The use of -t is deprecated, as the query-type is derived from the query result."
                ;;

      \?)       NAGIOS_SHORT_TEXT="invalid option: -$OPTARG"
                NAGIOS_LONG_TEXT="$(usage)"
                exit
                ;;

      \:)       NAGIOS_SHORT_TEXT="-$OPTARG requires an arguement"
                NAGIOS_LONG_TEXT="$(usage)"
                exit
                ;;
    esac
  done

  # check for missing parameters
  if [[ -z ${PROMETHEUS_SERVER} ]] ||
     [[ -z ${PROMETHEUS_QUERY} ]] ||
     [[ -z ${METRIC_NAME} ]] ||
     [[ -z ${WARNING_LEVEL} ]] ||
     [[ -z ${CRITICAL_LEVEL} ]]
  then
    NAGIOS_SHORT_TEXT='missing required option'
    NAGIOS_LONG_TEXT+="$(usage)"
    exit
  fi
  # Convert old syntax to new syntax
  # Warning
  if is_float "${WARNING_LEVEL}"; then
      if [[ "${COMPARISON_METHOD}" == "ge" ]]; then
          WARNING_LEVEL="0:$(echo ${WARNING_LEVEL} | jq '. - 1e-9')"
      elif [[ "${COMPARISON_METHOD}" == "eq" ]]; then
          WARNING_LEVEL="@${WARNING_LEVEL}:${WARNING_LEVEL}"
      elif [[ "${COMPARISON_METHOD}" == "ne" ]]; then
          WARNING_LEVEL="${WARNING_LEVEL}:${WARNING_LEVEL}"
      elif [[ "${COMPARISON_METHOD}" == "lt" ]]; then
          WARNING_LEVEL="${WARNING_LEVEL}:"
      elif [[ "${COMPARISON_METHOD}" == "le" ]]; then
          WARNING_LEVEL="$(echo ${WARNING_LEVEL} | jq '. + 1e-9'):"
      fi
  fi
  # Critical
  if is_float "${CRITICAL_LEVEL}"; then
      if [[ "${COMPARISON_METHOD}" == "ge" ]]; then
          CRITICAL_LEVEL="0:$(echo ${CRITICAL_LEVEL} | jq '. - 1e-9')"
      elif [[ "${COMPARISON_METHOD}" == "eq" ]]; then
          CRITICAL_LEVEL="@${CRITICAL_LEVEL}:${CRITICAL_LEVEL}"
      elif [[ "${COMPARISON_METHOD}" == "ne" ]]; then
          CRITICAL_LEVEL="${CRITICAL_LEVEL}:${CRITICAL_LEVEL}"
      elif [[ "${COMPARISON_METHOD}" == "lt" ]]; then
          CRITICAL_LEVEL="${CRITICAL_LEVEL}:"
      elif [[ "${COMPARISON_METHOD}" == "le" ]]; then
          CRITICAL_LEVEL="$(echo ${CRITICAL_LEVEL} | jq '. + 1e-9'):"
      fi
  fi
  # Decode new syntax
  # Warning
  WARNING_RANGE=$(decode_range ${WARNING_LEVEL})
  WARNING_LEVEL_LOW=$(echo "${WARNING_RANGE}" | cut -f1 -d' ')
  WARNING_LEVEL_HIGH=$(echo "${WARNING_RANGE}" | cut -f2 -d' ')
  WARNING_INVERTED=$(echo "${WARNING_RANGE}" | cut -f3 -d' ')
  # Critical
  CRITICAL_RANGE=$(decode_range ${CRITICAL_LEVEL})
  CRITICAL_LEVEL_LOW=$(echo "${CRITICAL_RANGE}" | cut -f1 -d' ')
  CRITICAL_LEVEL_HIGH=$(echo "${CRITICAL_RANGE}" | cut -f2 -d' ')
  CRITICAL_INVERTED=$(echo "${CRITICAL_RANGE}" | cut -f3 -d' ')
  # Bake our levels json
  LEVEL_JSON="{\"critical_low\": ${CRITICAL_LEVEL_LOW}, \"critical_high\": ${CRITICAL_LEVEL_HIGH}, \"warning_low\": ${WARNING_LEVEL_LOW}, \"warning_high\": ${WARNING_LEVEL_HIGH}}"
  # Sanity check critical and warning levels
  if [ ${COMPARISON_METHOD} != "ne" ]; then
      echo "${LEVEL_JSON}" | jq -e ".critical_low <= .critical_high" >/dev/null
      CRITICAL_LEVEL_OK=$?
      echo "${LEVEL_JSON}" | jq -e ".warning_low <= .warning_high" >/dev/null
      WARNING_LEVEL_OK=$?
      if [ ${CRITICAL_LEVEL_OK} -ne 0 ]; then
        NAGIOS_STATUS=UNKNOWN
        NAGIOS_SHORT_TEXT="invalid critical range: ${CRITICAL_LEVEL_LOW} <= ${CRITICAL_LEVEL_HIGH}"
        exit
      fi
      if [ ${WARNING_LEVEL_OK} -ne 0 ]; then
        NAGIOS_STATUS=UNKNOWN
        NAGIOS_SHORT_TEXT="invalid warning range: ${WARNING_LEVEL_LOW} <= ${WARNING_LEVEL_HIGH}"
        exit
      fi
  fi
}


function check_prometheus_server {
    # Check that the server URL responds
    curl --silent --head ${PROMETHEUS_SERVER}/api/v1/query > /dev/null
    SERVER_RESPONDED=$?
    if [ "${SERVER_RESPONDED}" -ne 0 ]; then
        NAGIOS_STATUS=UNKNOWN
        NAGIOS_SHORT_TEXT="no response from prometheus"
        exit
    fi

    # Check if the prometheus endpoint exists, and can be queried
    curl -s ${PROMETHEUS_SERVER}/api/v1/query?query=cafebabe | jq -e '.status == "success"' 1>/dev/null 2>/dev/null
    PROMETHEUS_OK=$?
    if [ "${PROMETHEUS_OK}" -ne 0 ]; then
        NAGIOS_STATUS=UNKNOWN
        NAGIOS_SHORT_TEXT="unable to query prometheus endpoint!"
        exit
    fi
}


function get_prometheus_raw_result {

  local _RESULT

  _RESULT=$(curl -sgG "${CURL_OPTS[@]}" --data-urlencode "query=${PROMETHEUS_QUERY}" "${PROMETHEUS_SERVER}/api/v1/query")
  printf '%s' "${_RESULT}"

}

function get_prometheus_scalar_result {

  local _RESULT

  _RESULT=$(echo $1 | jq -r '.[1]')

  # check result
  case "${_RESULT}" in
    +Inf) printf '%s' 'inf'
          ;;
    -Inf) printf '%s' '-inf'
          ;;
    *)    printf '%s' "${_RESULT}" # otherwise return as a string
          ;;
  esac
}

function get_prometheus_vector_value {

  local _RESULT

  # return the value of the first element of the vector
  _RESULT=$(echo $1 | jq -r '.[0].value?')
  printf '%s' "${_RESULT}"

}

function get_prometheus_vector_metric {

  local _RESULT

  # return the metric information of the first element of the vector
  _RESULT=$(echo $1 | jq -r '.[0].metric?' | xargs)
  printf '%s' "${_RESULT}"

}

if [[ "${BASH_SOURCE[0]}" == "${0}" ]]; then
    # Exit trigger
    #-------------
    # Function to be trapped on exit
    function on_exit {
        if [[ -n ${NAGIOS_LONG_TEXT} ]]; then
            printf '%s\n' "${NAGIOS_LONG_TEXT}"
        fi
        printf '%s - %s\n' ${NAGIOS_STATUS} "${NAGIOS_SHORT_TEXT}"
        # Indirect variable reference
        exit ${!NAGIOS_STATUS}
    }

    # Set up exit function
    trap on_exit EXIT TERM

    check_dependencies
    # process the cli options
    process_command_line "$@"

    # Check that we can communicate with the prometheus server
    check_prometheus_server

    # get the raw query from prometheus
    PROMETHEUS_RAW_RESPONSE="$( get_prometheus_raw_result )"

    PROMETHEUS_QUERY_TYPE=$(echo "${PROMETHEUS_RAW_RESPONSE}" | jq -r '.data.resultType')
    PROMETHEUS_RAW_RESULT=$(echo "${PROMETHEUS_RAW_RESPONSE}" | jq -r '.data.result')

    # extract the metric value from the raw prometheus result
    if [[ "${PROMETHEUS_QUERY_TYPE}" == "scalar" ]]; then
        PROMETHEUS_RESULT=$( get_prometheus_scalar_result "$PROMETHEUS_RAW_RESULT" )
        PROMETHEUS_METRIC=UNKNOWN
    else
        PROMETHEUS_VALUE=$( get_prometheus_vector_value "$PROMETHEUS_RAW_RESULT" )
        PROMETHEUS_RESULT=$( get_prometheus_scalar_result "$PROMETHEUS_VALUE" )
        PROMETHEUS_METRIC=$( get_prometheus_vector_metric "$PROMETHEUS_RAW_RESULT" )
    fi

    # check the value
    if is_float ${PROMETHEUS_RESULT}; then
      JSON=$(echo "${LEVEL_JSON} {\"value\": ${PROMETHEUS_RESULT}}" | jq -s add)
      # Evaluate critical and warning levels
      echo "${JSON}" | jq -e ".critical_low <= .value and .value <= .critical_high" >/dev/null
      IS_CRITICAL=$?
      echo "${JSON}" | jq -e ".warning_low <= .value and .value <= .warning_high" >/dev/null
      IS_WARNING=$?

      if [ ${IS_CRITICAL} -ne ${CRITICAL_INVERTED} ]; then
        NAGIOS_STATUS=CRITICAL
        NAGIOS_SHORT_TEXT="${METRIC_NAME} is ${PROMETHEUS_RESULT}"
      elif [ ${IS_WARNING} -ne ${WARNING_INVERTED} ]; then
        NAGIOS_STATUS=WARNING
        NAGIOS_SHORT_TEXT="${METRIC_NAME} is ${PROMETHEUS_RESULT}"
      else
        NAGIOS_STATUS=OK
        NAGIOS_SHORT_TEXT="${METRIC_NAME} is ${PROMETHEUS_RESULT}"
      fi
    else
      if [[ "${NAN_OK}" = "true" && "${PROMETHEUS_RESULT}" = "NaN" ]]; then
        NAGIOS_STATUS=OK
        NAGIOS_SHORT_TEXT="${METRIC_NAME} is ${PROMETHEUS_RESULT}"
      else
        NAGIOS_SHORT_TEXT="unable to parse prometheus response"
        NAGIOS_LONG_TEXT="${METRIC_NAME} is ${PROMETHEUS_RESULT}"
      fi
    fi
    if [[ "${NAGIOS_INFO}" = "true" ]]; then
        NAGIOS_SHORT_TEXT="${NAGIOS_SHORT_TEXT}: ${PROMETHEUS_METRIC}"
    fi
    if [[ "${PERFDATA}" = "true" ]]; then
        # Bake performance data
        PERF_DATA=""
        PERF_DATA+="query_result=${PROMETHEUS_RESULT};${WARNING_LEVEL};${CRITICAL_LEVEL};U;U"
        # PERF_DATA+=" "

        NAGIOS_SHORT_TEXT="${NAGIOS_SHORT_TEXT} | ${PERF_DATA}"
    fi

    exit
fi


src/usage.bash
function usage() {

  cat <<'EoL'

  check_prometheus_metric.sh - Nagios plugin for checking Prometheus metrics.

  Usage:
    check_prometheus_metric.sh -H HOST -q QUERY -w FLOAT[:FLOAT] -c FLOAT[:FLOAT]
                               -n NAME [-m METHOD] [-O] [-i] [-p]

  Options:
    -H HOST          URL of Prometheus host to query.
    -q QUERY         Prometheus query, in single quotes, that returns a float.
    -w FLOAT[:FLOAT] Warning level value (must be a float or nagios-interval).
    -c FLOAT[:FLOAT] Critical level value (must be a float or nagios-interval).
    -n NAME          A name for the metric being checked.
    -m METHOD        Comparison method, one of gt, ge, lt, le, eq, ne.
                     (Defaults to ge unless otherwise specified).
    -C CURL_OPTS     Additional flags to curl. Can be passed multiple times.
                     Options and option values must be passed separately.
                     e.g. -C --connect-timeout -C 10 -C --cacert -C /path/to/ca.crt
    -O               Accept NaN as an "OK" result.
    -i               Print the extra metric information into the Nagios message.
    -p               Add perfdata to check output.

  Examples:
    check_prometheus_metric -q 'up{job=\"job_name\"}' -w :1 -c :1
    # Check that job is up. If not, critical.

    check_prometheus_metric -q 'node_load1' -w :0.05 -c :0.1
    # Check load is below 0.05 (warning) and 0.1 (critical).

    check_prometheus_metric -q 'go_threads' -w 15:25 -c :
    # Check thread count is between 15-25, warning if outside this interval.

  Dependencies:
    Requires bash, curl, cut, echo, grep, jq and sed to be in $PATH.

EoL
}
